WalletWasabi/BitcoinCore/Mempool/MempoolMirror.cs

```diff
@@ -66,94 +56,69 @@ private void MempoolService_TransactionReceived(object? sender, SmartTransaction
 		AddTransactions(stx.Transaction);
 	}
 
-	internal int AddTransactions(params Transaction[] txs) => AddTransactions(txs as IEnumerable<Transaction>);
-
-	private int AddTransactions(IEnumerable<Transaction> txs)
+	internal int AddTransactions(params Transaction[] txs)
 	{
-		var added = 0;
 		lock (MempoolLock)
 		{
-			// Loop through the newly received transactions expect those that are already present in our mempool snapshot.
-			foreach (Transaction tx in txs.Where(x => !Mempool.ContainsKey(x.GetHash())).ToArray())
-			{
-				// Evict double spends.
-				foreach (TxIn txInput in tx.Inputs)
-				{
-					// tx spends a UTXO which some *other* transaction in the mempool *already* spends.
-					if (PrevOutsIndex.TryGetValue(txInput.PrevOut, out Transaction? mempoolTx))
-					{
-						// Remove the old transaction from our mempool snapshot.
-						RemoveTransactionNoLock(mempoolTx.GetHash());
-					}
-				}
-
-				AddTransactionNoLock(tx);
-				added++;
-			}
+			return Mempool.AddMissingTransactions(txs);
 		}
-		return added;
 	}
 
-	private void AddTransactionNoLock(Transaction tx)
+	private async Task<int> MirrorMempoolAsync(CancellationToken cancel)
 	{
-		Mempool.Add(tx.GetHash(), tx);
+		// Get all TXIDs in the up-to-date mempool.
+		uint256[]? newTxids = await Rpc.GetRawMempoolAsync(cancel).ConfigureAwait(false);
 
-		foreach (TxIn txInput in tx.Inputs)
-		{
-			PrevOutsIndex.Add(txInput.PrevOut, tx);
-		}
-	}
+		Mempool newMempool;
 
-	private void RemoveTransactionNoLock(uint256 txHash)
-	{
-		if (Mempool.Remove(txHash, out Transaction? transaction))
+		// TXIDs in the new mempool, but not in the old mempool (what we currently have).
+		IEnumerable<uint256> missingTxids;
+
+		lock (MempoolLock)
 		{
-			// Remove all UTXOs of the removed transaction from our helper index.
-			foreach (TxIn txInput in transaction.Inputs)
+			newMempool = Mempool.Clone();
+
+			ISet<uint256> oldTxids = Mempool.GetMempoolTxids();
+
```

Comment:
But if we clone the `Mempool` in the first place, wouldn't these return the same?

```
// Cloning beforehand
Assert.Equal(Mempool.GetMempoolTxids(), newMempool.GetMempoolTxids())
```

Reply:
It's not a mandatory change at all, but if we can hold the lock a bit less, then it might make sense to do it.
Of course, only if it's not breaking anything.

Reply:
What I mean is,
when we clone the `Mempool`, inside the lock `newMempool = Mempool.Clone();`

```
public Mempool Clone()
{
	return new Mempool()
	{
		TransactionIndex = new Dictionary<uint256, Transaction>(TransactionIndex),
		PrevOutsIndex = new Dictionary<OutPoint, Transaction>(PrevOutsIndex)
	};
}
```
we already have the existing (previous) `TransactionIndex` & `PrevOutsIndex`.

So I don't understand why we need to protect this line with the lock
`oldTxids = Mempool.GetMempoolTxids();`

when we could use the `newMempool` without the lock
`oldTxids = newMempool.GetMempoolTxids();`

I hope I make sense 😅

Reply:
Yeah, I had a feeling that it's only micro-optimization and it won't make a big difference. 
But it was just a suggestion, it wasn't a blocker or anything.